src/devices/sound/qsoundhle.cpp

```diff
@@ -180,70 +212,429 @@ WRITE8_MEMBER(qsound_hle_device::qsound_w)
+
+// Process a sample update
+void qsound_hle_device::state_normal_update()
+{
+	m_ready_flag = 0x80;
+	int32_t echo_input = 0;
+	int16_t echo_output;
+
+	// recalculate echo length
+	if(m_state == STATE_NORMAL2)
+		m_echo.length = m_echo.end_pos - DELAY_BASE_OFFSET2;
+	else
+		m_echo.length = m_echo.end_pos - DELAY_BASE_OFFSET;
+
+	m_echo.length = CLAMP(m_echo.length, 0, 1024);
+
+	// update PCM voices
+	for(int i=0; i<16; i++)
+		m_voice_output[i] = pcm_update(&m_voice[i], &echo_input);
+
+	// update ADPCM voices (one every third sample)
+	adpcm_update(m_state_counter % 3, m_state_counter / 3);
+
+	echo_output = echo(&m_echo,echo_input);
+
+	// now, we do the magic stuff
+	for(int ch=0; ch<2; ch++)
+	{
+		// Echo is output on the unfiltered component of the left channel and
+		// the filtered component of the right channel.
+		int32_t wet = (ch == 1) ? echo_output<<16 : 0;
+		int32_t dry = (ch == 0) ? echo_output<<16 : 0;
+		int32_t output = 0;
+
+		for(int i=0; i<19; i++)
+		{
+			uint16_t pan_index = m_voice_pan[i] + (ch * PAN_TABLE_CH_OFFSET);
+
+			// Apply different volume tables on the dry and wet inputs.
+			dry -= (m_voice_output[i] * (int16_t) read_dsp_rom(pan_index + PAN_TABLE_DRY))<<2;
+			wet -= (m_voice_output[i] * (int16_t) read_dsp_rom(pan_index + PAN_TABLE_WET))<<2;
+		}
+
+		// Apply FIR filter on 'wet' input
+		wet = fir(&m_filter[ch], wet >> 16);
+
+		// in mode 2, we do this on the 'dry' input too
+		if(m_state == STATE_NORMAL2)
+			dry = fir(&m_alt_filter[ch], dry >> 16);
+
+		// output goes through a delay line and attenuation
+		output = (delay(&m_wet[ch], wet) + delay(&m_dry[ch], dry))<<2;
+
+		// DSP round function
+		output = (output + 0x8000) & ~0xffff;
+		m_out[ch] = output >> 16;
+
+		if(m_delay_update)
+		{
+			delay_update(&m_wet[ch]);
+			delay_update(&m_dry[ch]);
+		}
+	}
+
+	m_delay_update = 0;
+
+	// after 6 samples, the next state is executed.
+	m_state_counter++;
+	if(m_state_counter > 5)
+	{
+		m_state_counter = 0;
+		m_state = m_next_state;
+	}
+}
+
+// Apply the FIR filter used as the Q1 transfer function
+int32_t qsound_hle_device::fir(struct qsound_fir *f, int16_t input)
+{
+	int32_t output = 0, tap = 0;
+
+	for(; tap < (f->tap_count-1); tap++)
+	{
+		output -= (f->taps[tap] * f->delay_line[f->delay_pos++])<<2;
+
+		if(f->delay_pos >= f->tap_count-1)
+			f->delay_pos = 0;
 	}
+
+	output -= (f->taps[tap] * input)<<2;
+
+	f->delay_line[f->delay_pos++] = input;
+	if(f->delay_pos >= f->tap_count-1)
+		f->delay_pos = 0;
+
+	return output;
+}
+
+// Apply delay line and component volume
+int32_t qsound_hle_device::delay(struct qsound_delay *d, int32_t input)
+{
+	int32_t output;
+
+	d->delay_line[d->write_pos++] = input>>16;
+	if(d->write_pos >= 51)
+		d->write_pos = 0;
+
+	output = d->delay_line[d->read_pos++]*d->volume;
+	if(d->read_pos >= 51)
+		d->read_pos = 0;
+
+	return output;
+}
+
+// Update the delay read position to match new delay length
+void qsound_hle_device::delay_update(struct qsound_delay *d)
```

Comment:
Things that operate entirely on an inner structure/class like this should be members of the inner structure/class, otherwise you're passing a useless `this` pointer for the outer class and need an explicit parameter for the inner structure/class.